contrib/runners/winrm_runner/winrm_runner/winrm_base.py

```diff
@@ -0,0 +1,301 @@
+
+    def _create_session(self):
+        # create the session
+        LOG.info("Connecting via WinRM to url: {}".format(self._winrm_url))
+        session = Session(self._winrm_url,
+                          auth=(self._username, self._password),
+                          transport=self._transport,
+                          server_cert_validation=self._server_cert_validation,
+                          operation_timeout_sec=self._timeout,
+                          read_timeout_sec=self._read_timeout)
+        return session
+
+    def _winrm_get_command_output(self, protocol, shell_id, command_id):
+        # NOTE: this is copied from pywinrm because it doesn't support
+        # timeouts
+        stdout_buffer, stderr_buffer = [], []
+        return_code = 0
+        command_done = False
+        start_time = time.time()
+        while not command_done:
+            # check if we need to timeout (StackStorm custom)
+            current_time = time.time()
+            elapsed_time = (current_time - start_time)
+            if self._timeout and (elapsed_time > self._timeout):
+                raise WinRmRunnerTimoutError(Response((b''.join(stdout_buffer),
+                                                       b''.join(stderr_buffer),
+                                                       WINRM_TIMEOUT_EXIT_CODE)))
+            # end stackstorm custom
+
+            try:
+                stdout, stderr, return_code, command_done = \
+                    protocol._raw_get_command_output(shell_id, command_id)
+                stdout_buffer.append(stdout)
+                stderr_buffer.append(stderr)
+            except WinRMOperationTimeoutError:
+                # this is an expected error when waiting for a long-running process,
+                # just silently retry
+                pass
+        return b''.join(stdout_buffer), b''.join(stderr_buffer), return_code
+
+    def _winrm_run_cmd(self, session, command, args=(), env=None, cwd=None):
+        # NOTE: this is copied from pywinrm because it doesn't support
+        # passing env and working_directory from the Session.run_cmd.
+        # It also doesn't support timeouts. All of these things have been
+        # added
+        shell_id = session.protocol.open_shell(env_vars=env,
+                                               working_directory=cwd)
+        command_id = session.protocol.run_command(shell_id, command, args)
+        # try/catch is for custom timeout handing (StackStorm custom)
+        try:
+            rs = Response(self._winrm_get_command_output(session.protocol,
+                                                         shell_id,
+                                                         command_id))
+            rs.timeout = False
+        except WinRmRunnerTimoutError as e:
+            rs = e.response
+            rs.timeout = True
+        # end stackstorm custom
+        session.protocol.cleanup_command(shell_id, command_id)
```

Comment:
Do we need to run those in `finally` to ensure those two functions are always called?

Reply:
The `pywinrm` library doesn't, but that doesn't mean it isn't a bad idea!